source/air/src/printer.rs

```diff
@@ -495,3 +495,182 @@ impl NodeWriter {
 pub(crate) fn node_to_string(node: &Node) -> String {
     NodeWriter::new().node_to_string_indent(&"".to_string(), node)
 }
+
+pub fn singular_query_to_node(vars: Vec<String>, enss: Vec<Expr>, reqs: Vec<Expr>) -> Node {
+    let pp = Printer::new(false);
+    let mut nodes = Vec::new();
+    let mut var_nodes = Vec::new();
+    nodes.push(str_to_node("singular-check-valid"));
+    let ens = pp.exprs_to_node(&Arc::new(enss));
+    let req = pp.exprs_to_node(&Arc::new(reqs));
+    for v in vars {
+        var_nodes.push(Node::Atom(v));
+    }
+    nodes.push(Node::List(var_nodes));
+    nodes.push(ens);
+    nodes.push(req);
+    Node::List(nodes)
+}
+
+use std::collections::HashMap;
+pub(crate) fn expr_to_singular(
+    expr: &Expr,
+    tmp_idx: &mut u32,
+    node_map: &mut HashMap<Node, String>,
+) -> String {
+    match &**expr {
+        ExprX::Const(Constant::Nat(n)) => n.to_string(),
+        ExprX::Var(x) => x.to_string(),
+        ExprX::Binary(BinaryOp::EuclideanMod, lhs, rhs) => {
+            // x % y ->  x - y*tmp
+            let pp = Printer::new(false);
+            let key = pp.expr_to_node(expr);
+            let value = node_map.get(&key);
+            let t = match value {
+                Some(tmp_var) => tmp_var.to_string(),
+                None => {
+                    let tmp_new = format!("tmp_{}", tmp_idx.to_string());
+                    *tmp_idx += 1;
+                    node_map.insert(key, tmp_new.to_string());
+                    tmp_new
+                }
+            };
+            let s1 = expr_to_singular(lhs, tmp_idx, node_map);
+            let s2 = expr_to_singular(rhs, tmp_idx, node_map);
+
+            format!("(({}) - (({})*({})))", s1, s2, t)
+        }
+        ExprX::Binary(BinaryOp::Eq, lhs, rhs) => {
+            let s1 = expr_to_singular(lhs, tmp_idx, node_map);
+            let s2 = expr_to_singular(rhs, tmp_idx, node_map);
+            format!("({}) - ({})", s1, s2)
+        }
+        ExprX::Multi(op, exprs) => {
+            let mut ss = vec![];
+            let sop = match op {
+                MultiOp::Add => ") + (",
```

Comment:
Are these parenthesis starting/ending sub-expressions? Who's opening the first and closing the second?